idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,230 @@
+    gradient_f,gradient_f_dic, gradient_c,gradient_c_dic, line_dic = get_dfds_dcds(model, theta_names, tee)
+    dsdp_dic, col  = get_dsdp(model, theta_names, theta, var_dic,tee)      
+    x_list = [x for x in col if x not in var_dic.keys()]
+    p_list = theta_names #This makes cov and fxxp have the same order 
+    fxxp = []
+    for i in p_list:
+        fxxp_tmp = 0
+        for j in x_list:
+            fxxp_tmp = fxxp_tmp + gradient_f_dic['d(f)/d('+j+')']*dsdp_dic['d('+j+')/d('+i+')']
+        fxxp.append(fxxp_tmp)
+    fxxp = np.array(fxxp)
+    # calculate error propagation df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)
+    # objective function: 
+    for cc in list(model.component_data_objects(Objective,
+                                            active=True,
+                                            descend_into=True)):
+        # save gradient of the objective with respect to only theta_names
+        if any(ele in str(cc.expr) for ele in theta_names):
+            gradient_f_theta = gradient_f[[int(line_dic[v]) -1 for v in theta_names]]
+    propagation_f = {}
+    # calculate error propagation of the objective fuction
+    if 'gradient_f_theta' in locals():
+        propagation_f['objective'] = np.dot(gradient_f_theta,np.dot(cov,np.transpose(gradient_f_theta))) + np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
+    else:
+        propagation_f['objective'] = np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
+    # constraints: 
+    # calculate error propagation dc/dp*cov_p*dc/dp + (dc/dx*dx/dp)*cov_p*(dc/dx*dx/dp)
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True))) 
+    if num_constraints > 0:
+        # save constraints into gradient_cc only if constraints includes theta 
+        # gradient_c[i] is a sparse matrix with column number, row number, and value
+        gradient_cc = []
+        for i in range(gradient_c.shape[0]):
+            if gradient_c[i,[0]] in line_dic.values():
+                gradient_cc.append(gradient_c[i])
+        gradient_cc = np.array(gradient_cc)
+        # save unique constraints numbers that contain theta, index starts from 1. 
+        constriant_number = list(set(gradient_cc[:,[1]].flatten().astype(int)))       
+        # convert sparse matrix to dense matrix with (value, (row number, column number))
+        I,J,V =  gradient_cc[:,[1]].flatten().astype(int)-1, gradient_cc[:,[0]].flatten().astype(int)-1,  gradient_cc[:,[2]].flatten()
+        gradient_cc = sparse.coo_matrix((V,(I,J))).todense()
+        gradient_dic = {}
+        propagation_c = {}
+        # calculate error propagation of constraints
+        for r in range(1,num_constraints+1):
+            if r in constriant_number:
+                cxxp = []
+                for i in p_list:
+                    cxxp_tmp = 0
```

Comment:
Please add more comments here.

Reply:
Should this be removed if you are resetting cxxp_tmp in the next loop?

Reply:
From [Line `cxxp = [ ]`](https://github.com/IDAES/idaes-pse/blob/0854fab43a0e50da7860682718bdd3864fbabd9f/idaes/apps/uncertainty_propagation/uncertainties.py#L175) to[ Line `cxxp = np.array(cxxp)` ](https://github.com/IDAES/idaes-pse/blob/0854fab43a0e50da7860682718bdd3864fbabd9f/idaes/apps/uncertainty_propagation/uncertainties.py#L187) are redundant codes. I removed them.